### PR TITLE
Expose new kind of origins

### PR DESCRIPTION
**Pull request by dbanck**
This PR allows schemas to introduce a target for the module source attribute. We achieve this by introducing a new field `Targets` to the `BodySchema`.

Example within terraform-schema: https://github.com/hashicorp/terraform-schema/pull/101

Required for https://github.com/hashicorp/terraform-ls/issues/558

### CODE REVIEW

File: decoder/reference_origins.go [hunk 1]
```diff
@@ -111,6 +111,14 @@ func (d *PathDecoder) referenceOriginsInBody(body hcl.Body, bodySchema *schema.B
 			}
 		}
 
+		if aSchema.IsDepKey && bodySchema.LinksTo != nil {
```
Comment (dbanck) [inline decoder/reference_origins.go:114]: It _feels_ like this is not guarded enough and might introduce `DirectOrigin`s at unwanted locations.

File: schema/body_schema.go [hunk 2]
```diff
@@ -31,13 +31,22 @@ type BodySchema struct {
 	// TargetableAs represents how else the body may be targeted
 	// if not by its declarable attributes or blocks.
 	TargetableAs Targetables
+
+	// LinksTo indicates whether the something inside the body is
+	// linking to another location
+	LinksTo *LinksTo
```
Comment (dbanck) [inline schema/body_schema.go:37]: I'm very open to better naming suggetions

File: decoder/reference_origins.go (hunk 1 above)
Comment (dbanck) [reply decoder/reference_origins.go:114]: Furthermore, all origins would link to the same location, and we don't support having multiple links inside a body.

File: reference/origin_direct.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/hashicorp/hcl/v2"
+)
+
+// DirectOrigin represents an origin which directly targets a file
+// and doesn't need a matching target
+type DirectOrigin struct {
+	// Range represents a range of a local traversal or an attribute
+	Range hcl.Range
+
+	// TargetPath represents what Path does the origin targets
```
Comment (radeksimko) [inline reference/origin_direct.go:14]:
```suggestion
	// TargetPath represents what (directory) Path does the origin targets
```
Nitpick - just to clarify this is not a file path

**Review by radeksimko — APPROVED** [review]
Thank you for all the tests and comments.

LGTM - I just left you some very minor in-line suggestions/comments, but none are blocking.

**Review by radeksimko — APPROVED** [review]
:shipit:

**Pull request merged**